### PR TITLE
657 Update doc-string of MaskedDiceLoss

### LINKED ISSUE #657: Incorrect results for MaskedDiceLoss when softmax=True
When the softmax parameter `softmax=True` is used in `MaskedDiceLoss`, the masked pixels are not excluded from the computation since the softmax is still applied to the zeroed-out pixels. Example:

```python
import math
import torch
from monai.losses import MaskedDiceLoss, DiceLoss

# shape: (batch, classes, height, width)
input_prob = torch.tensor([[[[0.1, 0.8]], [[0.9, 0.2]]]])  # [1, 2, 1, 2]
input_logits = torch.tensor([[[[math.log(0.1), math.log(0.8)]], [[math.log(0.9), math.log(0.2)]]]]) # [1, 2, 1, 2]
target = torch.tensor([[[[1, 0]]]])  # [1, 1, 1, 2]
mask = torch.tensor([[[[1, 0]]]])  # [1, 1, 1, 2]

dice_loss_prob = MaskedDiceLoss(to_onehot_y=True, softmax=False, reduction='none')
dice_loss_logits = MaskedDiceLoss(to_onehot_y=True, softmax=True, reduction='none')

# This yields the same result
dice_loss_prob(input_prob, target)  # tensor([[0.1579, 0.1429]])
dice_loss_logits(input_logits, target)  # tensor([[0.1579, 0.1429]])

# This does not (the first one is the correct result)
dice_loss_prob(input_prob, target, mask=mask)  # tensor([[1.0000, 0.0526]])
dice_loss_logits(input_logits, target, mask=mask)  # tensor([[0.3750, 0.2500]])
```

The problem is that the [zeroing happens before the softmax](https://github.com/Project-MONAI/MONAI/blob/master/monai/losses/dice.py#L173).

My solution would be to apply the masking before [this line](https://github.com/Project-MONAI/MONAI/blob/master/monai/losses/dice.py#L118). Actually, I don't know why there is a `MaskedDiceLoss` class in the first place. I think an additional (optional) argument to the forward function of `DiceLoss` would be sufficient. So, I propose:
- Remove the `MaskedDiceLoss` class.
- Add an additional (optional) argument to the `DiceLoss` class for the mask.
- Add the example above to the tests.

If this is fine with you, I could make a PR :-)

Bug is present in commit 11fc31dc1b6b37cf3ed6a3a89e39739629cd3147

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #657 .

### Description
This PR updated the doc-string of `MaskedDiceLoss` to remove confusion.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/losses/dice.py [hunk 1]
```diff
@@ -147,7 +147,9 @@ def forward(self, input: torch.Tensor, target: torch.Tensor, smooth: float = 1e-
 
 class MaskedDiceLoss(DiceLoss):
     """
-    Same as DiceLoss, but accepts a binary mask ([0,1]) indicating a region over which to compute the dice.
+    Add an additional `masking` progress before `DiceLoss`, accept a binary mask ([0, 1]) indicating a region,
```
Comment (ericspod) [inline monai/losses/dice.py:150]: "`masking` progress" -> "`masking` process" ?

File: monai/losses/dice.py [hunk 2]
```diff
@@ -147,7 +147,9 @@ def forward(self, input: torch.Tensor, target: torch.Tensor, smooth: float = 1e-
...
+    `input` and `target` will be masked by the region: region with mask `1` will keep the original value,
+    region with `0` mask will be converted to `0`. Then feed `input` and `target` to normal `DiceLoss` computation.
```
Comment (ericspod) [inline monai/losses/dice.py:152]: I would append something like "This has the effect of ensuring only the masked region contributes to the loss computation and hence gradient calculation." Do we have some motivating use case where this is useful? It would help to explain why one would use a particular loss function over another.

File: monai/losses/dice.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/losses/dice.py:150]: updated in the latest commit.
Thanks.

File: monai/losses/dice.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/losses/dice.py:152]: Actually, I think this function was added as kind of "attention" mechanism, it still computes the unmasked region as `0` values, not only the masked region region contributes to the loss computation.
What do you think?
Thanks.

Comment (ericspod) [reply monai/losses/dice.py:152]: If unmasked regions are 0 I wouldn't think they really contributing to the gradient. Is the mask supposed to be learnable? I don't see what the use case would be of this still.

Comment (Nic-Ma) [reply monai/losses/dice.py:152]: I double-checked the source code, I think you are right. Updated according to your comments.
And I think maybe you can check the original bug: https://github.com/Project-MONAI/MONAI/issues/657?, @wyli may know more about the usage example.

Thanks.

Comment (ericspod) [reply monai/losses/dice.py:152]: I think @wyli should check this then, the usage described in the bug report is along the lines of what I was thinking in which case we haven't fixed the bug. If this usage isn't the intended one then we need to clarification on what the attention mechanism is.

Comment (Nic-Ma) [reply monai/losses/dice.py:152]: Sure, @wyli , could you please help share some comments?
Thanks.

Comment (wyli) [reply monai/losses/dice.py:152]: yes my understanding is that the mask tensor could be learnable, a similar approach ("voxel-level difficulty-aware attention") is used in this recent IJCV paper: https://link.springer.com/article/10.1007/s11263-020-01321-2

Comment (Nic-Ma) [reply monai/losses/dice.py:152]: @wyli , so you suggest changing the existing `MaskedDiceLoss` instead of just updating its doc-string?
If yes, maybe we should discuss this with @myron ?
Thanks.

Comment (wyli) [reply monai/losses/dice.py:152]: i think updating the docstring would be good enough, but perhaps @myron could review the issue and this PR

Comment (Nic-Ma) [reply monai/losses/dice.py:152]: OK, sure, let me invite @myron for a review here.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @myron , could you please help review this PR as it modified the doc-string of your original code.
Thanks.

**Comment by myron** [comment]
the changes look good.

**Comment by Nic-Ma** [comment]
Hi @ericspod , could you please help take a look? @myron is the original author of `MaskedDiceLoss`.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod , I resolved conflicts and merged master.
Do you have any other comments?

Thanks.

**Comment by ericspod** [comment]
I'm happy with it then to be merged.
